### PR TITLE
Add GitHub workflow for E2EV2

### PR DESCRIPTION
**Pull request by OliverMKing**
# Description

Adds E2EV2 to validating GitHub workflows

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Tested on my fork. 

https://github.com/OliverMKing/aks-app-routing-operator/actions/runs/6237297390

You can see the new stuff (E2EV2) working here. e2e (the old one) is failing because I didn't set up full Azure OAuth permissions on my fork. It should work in this repo.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/workflows/e2ev2.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+name: E2E Version 2 Test
```
Comment (jaiveerk) [inline .github/workflows/e2ev2.yaml:1]: can we rename to generate e2e matrix or something similar?

File: docs/e2e.md [hunk 2]
```diff
@@ -51,7 +54,14 @@ Tests are defined in [/testing/e2e/suites/](../testing/e2e/suites/). Add any new
 
 ## GitHub Runner
 
-todo: will fill in details when I add GitHub workflows to run new e2e
+We use GitHub workflows to run and require passing E2E tests on every PR. [.github/workflows/ok-to-test.yaml](../.github/workflows/ok-to-test.yaml), [.github/workflows/pr-validate.yaml](../.github/workflows/pr-validate.yaml), [.github/workflows/pr-validate-fork.yaml](../.github/workflows/pr-validate-fork.yaml), [.github/workflows/e2ev2.yaml](../.github/workflows/e2ev2.yaml), and [.github/workflows/e2ev2-provision-test.yaml](../.github/workflows/e2ev2-provision-test.yaml) chain together to provide this.
```
Comment (jaiveerk) [inline docs/e2e.md:57]: beyond "chain together" can you add some description about what each one does?

File: .github/workflows/e2ev2.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      - uses: actions/setup-go@v4
+        with:
+          go-version: '~1.20.3'
+          cache-dependency-path: "**/*.sum"
+
+      - run: |
+          cd testing/e2e
+          go run ./main.go matrix
+        shell: bash
+        id: matrix
```
Comment (jaiveerk) [inline .github/workflows/e2ev2.yaml:34]: I imagine it's a github glitch that the formatting isn't applying to the keys here and not an indentation issue?

**Review by jaiveerk — CHANGES_REQUESTED** [review]
minor changes but looks good, really like the use of multiple workflows

File: .github/workflows/e2ev2-provision-test.yaml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        with:
+          client-id: ${{ secrets.AZURE_CLIENT_ID }}
+          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
+          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
+
+      - name: Provision Infrastructure
+        shell: bash
+        run: (cd testing/e2e && go run ./main.go infra --subscription="${{ secrets.AZURE_SUBSCRIPTION_ID }}" --tenant="${{ secrets.AZURE_TENANT_ID }}" --names="${{ inputs.name }}" --infra-file="./infra.json")
+        if:
+          (github.event_name == 'repository_dispatch' &&
```
[the remark this answers is not in the document]
Comment (OliverMKing) [reply .github/workflows/e2ev2-provision-test.yaml:42]: done

File: .github/workflows/e2ev2.yaml (hunk 1 above)
Comment (OliverMKing) [reply .github/workflows/e2ev2.yaml]: done

File: docs/e2e.md (hunk 2 above)
Comment (OliverMKing) [reply docs/e2e.md]: done

File: .github/workflows/e2ev2.yaml (hunk 3 above)
Comment (OliverMKing) [reply .github/workflows/e2ev2.yaml]: looks fine to me and works

**Review by tbarnes94 — APPROVED** [review]

File: .github/workflows/e2ev2-matrix.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+      - uses: actions/setup-go@v4
+        with:
+          go-version: '~1.20.3'
+          cache-dependency-path: "**/*.sum"
+
+      - run: |
+          cd testing/e2e
+          go run ./main.go matrix
+        shell: bash
```
Comment (davidgamero) [inline .github/workflows/e2ev2-matrix.yaml:33]: i also see a weird formatting here, but idk

File: .github/workflows/e2ev2-matrix.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        id: matrix
+        if:
```
Comment (davidgamero) [inline .github/workflows/e2ev2-matrix.yaml:35]: i also see some formatting strangeness here
image

Comment (OliverMKing) [reply .github/workflows/e2ev2-matrix.yaml:35]: what is strange about this? nothing looks weird to me

Comment (OliverMKing) [reply .github/workflows/e2ev2-matrix.yaml:35]: it's been tested and working fine. wouldn't worry about weird syntax

Comment (jaiveerk) [reply .github/workflows/e2ev2-matrix.yaml:35]: the keys are the same color as the value - but if it works fine it's probably a github ui glitch

**Review by jaiveerk — APPROVED** [review]

File: .github/workflows/e2ev2-matrix.yaml (hunk 6 above)
Comment (OliverMKing) [reply .github/workflows/e2ev2-matrix.yaml:35]: github sev 1

**Review by bfoley13 — APPROVED** [review]
looks good

**Review by davidgamero — APPROVED** [review]
lgtm

**Pull request merged**